File: AdamOptimizer/AdamOptimizer/AdamOptimizer.cs
using System;

namespace AdamOptimizer
{
    /// <summary>
    /// 一つの重み値用のADAMオプティマイザー
    /// </summary>
    class AdamOptimizer
    {
        const double _epsilon = 0.00000001;
        readonly double _alpha;
        readonly double _beta1;
        readonly double _beta2;
        int _t = 0;
        double _m = 0;
        double _v = 0;

        public AdamOptimizer(double alpha = 0.001, double beta1 = 0.9, double beta2 = 0.999)
        {
            _alpha = alpha;
            _beta1 = beta1;
            _beta2 = beta2;
        }

        public double newW(   // 新しい重み値
            double w,   // 変更対象の重み値の現在の値
            double g)   // 重み値の微分値
        {
            _m = _beta1 * _m + (1 - _beta1) * g;
            _v = _beta2 * _v + (1 - _beta2) * g * g;

            var mHat = _m / (1 - Math.Pow(_beta1, _t));
            var vHat = _v / (1 - Math.Pow(_beta2, _t));
            var newW = w - _alpha * mHat / Math.Sqrt(vHat + _epsilon);

            _t++;

            return newW;
        }
    }
}

File: DeepLearningFromZeroKot/src/word2vec/AdamOptimizer.kt
package word2vec

import kotlin.math.pow
import kotlin.math.sqrt

// http://arxiv.org/abs/1412.6980v8
// ADAMオプティマイザー
class AdamOptimizer(
        paramCount: Int,    // 重み値の個数
        private val learningRate: Float = 0.001f,
        private val beta1: Float = 0.9f,
        private val beta2: Float = 0.999f) {
    private var count: Int = 0
    private val m = MutableList<Float>(paramCount) { 0f }
    private val v = MutableList<Float>(paramCount) { 0f }

    fun update(
            params: Array<Float>,  // 更新対象の重み値のリスト
            grads: Array<Float>) { // 重み値の傾きのリスト

        count += 1
        val lr = learningRate * sqrt(1f - beta2.pow(count)) / (1f - beta1.pow(count))

        params.indices.forEach { i ->
            m[i] += (1f - beta1) * (grads[i] - m[i])
            v[i] += (1f - beta2) * (grads[i].pow(2) - v[i])

            params[i] -= lr * m[i] / (sqrt(v[i]) + 1e-7f)
        }
    }
}
